src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicBuffers.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Microsoft.Quic;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal;
+
+/// <summary>
+/// Helper class to convert managed data into QUIC_BUFFER* consumable by MsQuic.
+/// It also allows reuse with repeated Reset/Initialize calls, e.g. new, Initialize, (use), Reset, Initialize, (use), Reset, Initialize, (use), Dispose.
+/// Note that since this is struct and there's no finalizer, Dispose must be always called to release the unmanaged memory allocated by this struct.
+/// </summary>
+internal unsafe struct MsQuicBuffers : IDisposable
+{
+    // ArrayPool only for MsQuicBuffers which always clears arrays upon return.
+    private static readonly ArrayPool<MemoryHandle> s_pool = ArrayPool<MemoryHandle>.Create();
```

Comment:
Do you have any performance numbers that suggest that the local instance of the ArrayPool works well here?

We have done a lot of performance tuning on the shared ArrayPool. The local ArrayPools (the kind you create using the `Create` method like here) received very little performance tuning.

Reply:
I've removed the pooling, it's taken from the original code/usage and not based on any measurements.